src/protocols/memory/ddr4.stanza

```diff
@@ -0,0 +1,453 @@
+
+    0.0 +/-    1.0e-12   ; Intra-pair skew match of CK is +/- 1 ps
+    0.0 +/-    1.0e-12   ; Intra-pair skew match of DQS is +/- 1 ps
+  -85.0e-12 +935.0e-12   ; Skew match between CK.P and DQS.P -85 ps to 935 ps
+    0.0 +/-    3.5e-12   ; Skew match between DQ and DQS.P within a byte lane is +/- 3.5 ps
+    0.0 +/-   20.0e-12   ; Skew match between A[17] A[13:0] and CK.P +/- 20 ps
+    0.0 +/-   20.0e-12   ; Skew match between RAS_n, CAS_n, WE_n, and CK.P is +/- 20 ps
+    0.0 +/-   20.0e-12   ; Skew match between BA[1:0], BG[1:0] and CK.P is +/- 20 ps
+    0.0 +/-   20.0e-12   ; Skew match between CMD/ADDR/CTRL within a channel is +/- 20 ps
+    0.0 +/-   20.0e-12   ; Skew match between ACT_n, CKE, CS_n, ODT, PAR and CK.P is +/- 20 ps
+
+The `constrain` function for this type expects two compatible `ddr4-b` types.
+    ck-rs:DifferentialRoutingStructure
+    dqs-rs:DifferentialRoutingStructure
+    dq-rs:RoutingStructure
+    rs:RoutingStructure ) -> DDR4-Constraint:
+
+<DOC>
+public defstruct DDR4-Constraint <: SI-Constraint :
+  doc: \<DOC>
+  Channel Bit Width
+  <DOC>
+  width:DDR4-width
+  doc: \<DOC>
+  Topology of memory channel
+  <DOC>
+  topology:DDR4-topology
+  doc: \<DOC>
+  Channel Rank
+  <DOC>
+  rank:DDR4-rank
+    doc: \<DOC>
+  CK Intra-pair Timing Skew Constraint in Seconds
+  <DOC>
+  skew-ck:Toleranced with :
+    default => 0.0 +/- 1.00e-12
+  doc: \<DOC>
+  DQS Intra-pair Timing Skew Constraint in Seconds
+  <DOC>
+  skew-dqs:Toleranced with :
+    default => 0.0 +/- 1.00e-12
+  doc: \<DOC>
+  CK.P to DQS.P Inter-signal Timing Skew Constraint in Seconds
+  <DOC>
+  skew-ck-dqs:Toleranced with :
+    default => min-typ-max(-85.0e-12 0.0 935.0e-12)
+  doc: \<DOC>
+  DQ to DQS Inter-pair Timing Skew Constraint in Seconds
+  <DOC>
+  skew-dq-dqs:Toleranced with :
+    default => 0.0 +/- 3.50e-12
+  doc: \<DOC>
+  DM_n to DQS.P Timing Skew Constraint in Seconds
+  <DOC>
+  skew-dm-dqs:Toleranced with:
+    default => 0.0 +/- 3.50e-12 ; Same as DQ timing
+  doc: \<DOC>
+  CMD/ADDR/CTRL to CK Inter-pair Timing Skew Constraint in Seconds
+  <DOC>
+  skew-cmd-addr-ctrl-ck:Toleranced with :
+    default => 0.0 +/- 20.0e-12
+  doc: \<DOC>
+  CMD, ADDR, CTRL (w/o CK) Intra-group Timing Skew Constraint in Seconds
+  <DOC>
+  skew-cmd-addr-ctrl:Toleranced  with:
+    default =>  0.0 +/- 10.0e-12
+  doc: \<DOC>
+  Diff-Pair Max Loss Limit Constraint in dB
+  <DOC>
+  loss:Double with:
+    ensure => ensure-positive!,
+    default => 5.0
+  doc: \<DOC>
+  Differential Routing Structure for CK  [90.0 +/- (5 %)]
+  <DOC>
+  diff-ck-rs:DifferentialRoutingStructure
+  doc: \<DOC>
+  Differential Routing Structure for DQS [100.0 +/- (5 %)]
+  <DOC>
+  diff-dqs-rs:DifferentialRoutingStructure
+  doc: \<DOC>
+  Single-Ended Routing Structure for each DQ signal [50.0 +/- (5 %)]
+  <DOC>
+  se-dq-rs:RoutingStructure
+  doc: \<DOC>
+  Single-Ended Routing Structure for all other SE signal [45.0 +/- (5 %)]
+  <DOC>
+  se-rs:RoutingStructure
```

Comment:
These are the routing structures not the direct impedances. The `DDR4-Impedance` struct would be used by user code because it would drive board-specific values.